File: sdks/dotnet/src/Hippo.Sdk/Exceptions.cs
namespace Hippo.Sdk;

/// <summary>Base exception for all Hippo API errors.</summary>
public class HippoException : Exception
{
    public int StatusCode { get; }

    public HippoException(string message, int statusCode)
        : base(message)
    {
        StatusCode = statusCode;
    }

    public HippoException(string message, int statusCode, Exception innerException)
        : base(message, innerException)
    {
        StatusCode = statusCode;
    }
}

/// <summary>Thrown when the server returns 401 Unauthorized.</summary>
public class AuthenticationException : HippoException
{
    public AuthenticationException(string message)
        : base(message, 401) { }
}

/// <summary>Thrown when the server returns 403 Forbidden.</summary>
public class ForbiddenException : HippoException
{
    public ForbiddenException(string message)
        : base(message, 403) { }
}

/// <summary>Thrown when the server returns 429 Too Many Requests.</summary>
public class RateLimitException : HippoException
{
    /// <summary>Parsed value of the Retry-After header, if present.</summary>
    public TimeSpan? RetryAfter { get; }

    public RateLimitException(string message, TimeSpan? retryAfter = null)
        : base(message, 429)
    {
        RetryAfter = retryAfter;
    }
}

File: sdks/python/hippo/exceptions.py
from __future__ import annotations


class HippoError(Exception):
    """Base exception for Hippo SDK errors."""

    def __init__(self, message: str, status_code: int | None = None, body: str | None = None) -> None:
        self.status_code = status_code
        self.body = body
        super().__init__(message)


class AuthenticationError(HippoError):
    """Raised on HTTP 401 — invalid or missing API key."""


class ForbiddenError(HippoError):
    """Raised on HTTP 403 — insufficient permissions."""


class RateLimitError(HippoError):
    """Raised on HTTP 429 — too many requests."""

    def __init__(
        self,
        message: str,
        status_code: int | None = None,
        body: str | None = None,
        retry_after: float | None = None,
    ) -> None:
        super().__init__(message, status_code=status_code, body=body)
        self.retry_after = retry_after
